includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+
+		$job_post = array(
+			'timestamp'     => $timestamp,
+			'action'        => $action,
+			'action_hashed' => md5( $action ),
+			'instance'      => md5( maybe_serialize( $args['args'] ) ),
+			'args'          => maybe_serialize( $args['args'] ),
+			'last_modified' => current_time( 'mysql', true ),
+		);
+
+		if ( isset( $args['schedule'] ) && ! empty( $args['schedule'] ) ) {
+			$job_post['schedule'] = $args['schedule'];
+		}
+
+		if ( isset( $args['interval'] ) && ! empty( $args['interval'] ) && is_numeric( $args['interval'] ) ) {
+			$job_post['interval'] = (int) $args['interval'];
+		}
+
+		// Create the post, or update an existing entry to run again in the future
+		if ( is_int( $update_id ) && $update_id > 0 ) {
+			$wpdb->update( $this->get_table_name(), $job_post, array( 'ID' => $update_id, ) );
+		} else {
+			$job_post['created'] = current_time( 'mysql', true );
+
+			$wpdb->insert( $this->get_table_name(), $job_post );
+		}
+
+		// Delete internal cache
+		// Should only be skipped during bulk operations
+		if ( $flush_cache ) {
+			$this->flush_internal_caches();
+		}
+	}
+
+	/**
+	 * Mark an event's entry as completed
+	 *
+	 * Completed entries will be cleaned up by an internal job
+	 *
+	 * @param int    $timestamp    Unix timestamp event executes at
+	 * @param string $action       Name of action used when the event is registered (unhashed)
+	 * @param string $instance     md5 hash of the event's arguments array, which Core uses to index the `cron` option
+	 * @param bool   $flush_cache  Whether or not to flush internal caches after creating/updating the event
+	 * @return bool
+	 */
+	public function mark_job_completed( $timestamp, $action, $instance, $flush_cache = true ) {
+		$job_id = $this->job_exists( $timestamp, $action, $instance, true );
+
+		if ( ! $job_id ) {
+			return false;
+		}
+
+		return $this->mark_job_record_completed( $job_id, $flush_cache );
+	}
+
+	/**
+	 * Set a job post to the "completed" status
+	 *
+	 * @param int $job_id        ID of job's record
+	 * @param bool $flush_cache  Whether or not to flush internal caches after creating/updating the event
+	 * @return bool
+	 */
+	public function mark_job_record_completed( $job_id, $flush_cache = true ) {
+		global $wpdb;
+
+		$updates = array(
+			'status'   => self::STATUS_COMPLETED,
+			'instance' => wp_rand( 1000000, 999999999 ), // Breaks unique constraint, and can be recreated from entry's remaining data
+		);
+
+		$success = $wpdb->update( $this->get_table_name(), $updates, array( 'ID' => $job_id, ) );
+
+		// Delete internal cache
+		// Should only be skipped during bulk operations
+		if ( $flush_cache ) {
+			$this->flush_internal_caches();
+		}
+
+		return (bool) $success;
+	}
+
+	/**
+	 * Compare two arrays and return collapsed representation of their differences
+	 *
+	 * @param array $new New cron array
+	 * @param array $old Old cron array
+	 *
+	 * @return array
+	 */
+	private function find_cron_array_differences( $new, $old ) {
```

Comment:
Minor: technically new and old are not accurate names since we swap the order to find deletes.